node-data/src/message.rs

```diff
@@ -968,23 +1001,119 @@ impl<M: Clone> AsyncQueue<M> {
     }
 }
 
+pub trait ConsensusMessage {
+    fn signable(&self) -> Vec<u8>;
+    fn header(&self) -> &ConsensusHeader;
+    fn header_mut(&mut self) -> &mut ConsensusHeader;
+
+    fn verify_signature(&self) -> Result<(), dusk_bls12_381_sign::Error> {
+        let signature = self.header().signature.inner();
+        let sig = dusk_bls12_381_sign::Signature::from_bytes(signature)?;
+        let pk =
+            dusk_bls12_381_sign::APK::from(self.header().pubkey_bls.inner());
+        let msg = self.signable();
+        pk.verify(&sig, &msg)
+    }
+
+    fn sign(
+        &mut self,
+        sk: &dusk_bls12_381_sign::SecretKey,
+        pk: &dusk_bls12_381_sign::PublicKey,
+    ) {
+        let msg = self.signable();
+        let signature = sk.sign(pk, &msg).to_bytes();
+        self.header_mut().signature = signature.into();
+    }
+}
+
+impl ConsensusMessage for Validation {
+    fn signable(&self) -> Vec<u8> {
+        let mut signable = self.header.signable();
+        signable.extend_from_slice(&[ConsensusMsgType::Validation as u8]);
+        self.vote
+            .write(&mut signable)
+            .expect("Writing to vec should succeed");
+        signable
+    }
+    fn header(&self) -> &ConsensusHeader {
+        &self.header
+    }
+    fn header_mut(&mut self) -> &mut ConsensusHeader {
+        &mut self.header
+    }
+}
+
+impl ConsensusMessage for Ratification {
+    fn signable(&self) -> Vec<u8> {
+        let mut signable = self.header.signable();
+        signable.extend_from_slice(&[ConsensusMsgType::Ratification as u8]);
+        self.vote
+            .write(&mut signable)
+            .expect("Writing to vec should succeed");
+        signable
+    }
+    fn header(&self) -> &ConsensusHeader {
+        &self.header
+    }
+    fn header_mut(&mut self) -> &mut ConsensusHeader {
+        &mut self.header
+    }
+}
+
+impl ConsensusMessage for Quorum {
+    fn signable(&self) -> Vec<u8> {
+        let mut signable = self.header.signable();
+        signable.extend_from_slice(&[ConsensusMsgType::Quorum as u8]);
+        self.vote
+            .write(&mut signable)
+            .expect("Writing to vec should succeed");
+        signable
+    }
+    fn header(&self) -> &ConsensusHeader {
+        &self.header
+    }
+    fn header_mut(&mut self) -> &mut ConsensusHeader {
+        &mut self.header
+    }
+}
+
+impl ConsensusMessage for Candidate {
+    fn signable(&self) -> Vec<u8> {
+        let mut signable = self.header.signable();
+        signable.extend_from_slice(&[ConsensusMsgType::Candidate as u8]);
```

Comment:
why is this needed?

Reply:
Every consensus message is verified at the beginning of the phase. I agree with you that this can be useless, since the block is already signed, but I would investigate a bit more before remove it

Reply:
Agreed